### PR TITLE
feat: upgrade to the OCaml 5.1.1 branch

### LINKED ISSUE #922: OCaml 5.1.1 support
https://discuss.ocaml.org/t/ocaml-5-1-1-incoming-breaking-change-in-the-marshal-module/13457

Melange is one of the very few packages affected by this.

### PR DESCRIPTION
**Pull request by anmonteiro**
fixes #922

### CODE REVIEW

**Comment by anmonteiro** [comment]
opened https://github.com/ocaml/opam-repository/pull/24845 to be able to disable and compression and fix the opam tests.

**Comment by anmonteiro** [comment]
I tried a few things but it looks like we'll need some kind of support from the opam-repository upstream for our OPAM pipeline to pass...

File: .github/workflows/opam-build.yml [hunk 1]
```diff
@@ -46,10 +46,15 @@ jobs:
         run: |
           env HOMEBREW_NO_AUTO_UPDATE=1 brew install tree
 
+      - name: Uninstall zstd?
+        run: sudo apt-get remove --auto-remove zstd
```
Comment (kit-ty-kate) [inline .github/workflows/opam-build.yml:50]:
```suggestion
        run: sudo apt-get remove --auto-remove libzstd-dev
```

Comment (kit-ty-kate) [reply .github/workflows/opam-build.yml:50]: Actually, the reason this step seems required seems dodgy. I believe a change in merlin is required instead. And in every opam packages that read `.cmi` files for that matter. That’s probably going to push the release a little.

**Comment by Octachron** [comment]
The error is that the vendored compilerlibs library is missing the compression module.

**Comment by anmonteiro** [comment]
> The root error seems to be that the vendored compilerlibs library is missing the `Compression` module.

Why is that an error? We don't vendor a lot of the unnecessary modules upstream, and we'd rather not duplicate `Compression`. 

Rather, we'd like to rely on the ability to turn off compression in CI, which is why I opened https://github.com/ocaml/opam-repository/pull/24845.

As you can see, the Nix tests are perfectly fine, because we don't include `libzstd` and the OCaml compiler configuration script detects that.

File: .github/workflows/opam-build.yml (hunk 1 above)
Comment (anmonteiro) [reply .github/workflows/opam-build.yml]: I just tried the code suggestion but looks like Dune itself needs lzstd? that seems weird.

**Comment by anmonteiro** [comment]
Now the `melange-playground` tests are failing – looks like we might need jsoo support for this release too:

> # +  Fatal error: exception Failure("caml_zstd_initialize not implemented")

**Comment by Octachron** [comment]
Indeed, `js_of_ocaml` need to implement the new primitve see https://github.com/ocsigen/js_of_ocaml/issues/1534 .

I missed the fact that you were using the compiler-libs to compiled your vendored file, so everything should work without issues. Which means that I don't understand why you need to disable compression for your tests?

**Comment by anmonteiro** [comment]
> Which means that I don't understand why you need to disable compression for your tests?

We use Merlin in our tests, which doesn’t seem to be able to read the compressed artifacts for 5.1 (perhaps because it’s been removed from the stdlib — Merlin might need to be updated as well).

**Comment by kit-ty-kate** [comment]
> Merlin might need to be updated as well

Yes I mentionned that in my above review. This is being worked on here: https://github.com/ocaml/merlin/pull/1714

**Comment by anmonteiro** [comment]
Thank you both for the assistance.

**Pull request merged**